transaction.go

```diff
@@ -0,0 +1,489 @@
+// call to Get. IsPopulated returns false if io.Address is nil
+func (io FactoidTransactionIO) IsPopulated() bool {
+	return io.Address != nil
+}
+
+// IsPopulated returns true if s has already been successfully populated by a
+// call to Get. IsPopulated returns false if s.SignatureBlock or s.ReedeemCondition are nil
+func (s FactoidTransactionSignature) IsPopulated() bool {
+	return s.SignatureBlock != nil &&
+		s.ReedeemCondition != nil
+}
+
+// Valid returns if the inputs of the factoid transaction are properly signed by the redeem conditions.
+// It will also validate the total inputs is greater than the total outputs.
+func (s *FactoidTransaction) Valid() bool {
+	if !s.IsPopulated() {
+		return false
+	}
+
+	// Validate amounts
+	if s.TotalFCTInputs() < s.TotalFCTOutputs()+s.TotalECOutput() {
+		return false
+	}
+
+	// Validate signatures
+	if len(s.FCTInputs) != len(s.Signatures) {
+		return false
+	}
+
+	msg, err := s.MarshalLedgerBinary()
+	if err != nil {
+		return false
+	}
+
+	for i := range s.FCTInputs {
+		expAddr, err := s.Signatures[i].ReedeemCondition.Address()
+		if err != nil {
+			return false
+		}
+		// RCD should match the input
+		if *expAddr != *s.FCTInputs[i].Address {
+			return false
+		}
+
+		if !s.Signatures[i].Validate(msg) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s *FactoidTransaction) TotalFCTInputs() (total uint64) {
+	return FactoidTransactionIOs(s.FCTInputs).TotalAmount()
+}
+
+func (s *FactoidTransaction) TotalFCTOutputs() (total uint64) {
+	return FactoidTransactionIOs(s.FCTOutputs).TotalAmount()
+}
+
+// TotalECOutput is delimated in factoishis
+func (s *FactoidTransaction) TotalECOutput() (total uint64) {
+	return FactoidTransactionIOs(s.ECOutputs).TotalAmount()
+}
+
+func (s FactoidTransactionIOs) TotalAmount() (total uint64) {
+	for _, io := range s {
+		total += io.Amount
+	}
+	return
+}
+
+func (s FactoidTransactionSignature) Validate(msg Bytes) bool {
+	return s.ReedeemCondition.Validate(msg, s.SignatureBlock)
+}
+
+// Get queries factomd for the entry corresponding to f.TransactionID, which must be not
+// nil. After a successful call all inputs, outputs, and the header will be populated
+func (f *FactoidTransaction) Get(c *Client) error {
+	// TODO: Test this functionality
+	// If the TransactionID is nil then we have nothing to query for.
+	if f.TransactionID == nil {
+		return fmt.Errorf("txid is nil")
+	}
+	// If the Transaction is already populated then there is nothing to do. If
+	// the Hash is nil, we cannot populate it anyway.
+	if f.IsPopulated() {
+		return nil
+	}
+	requestTxId := f.TransactionID
+
+	params := struct {
+		Hash *Bytes32 `json:"hash"`
+	}{Hash: f.TransactionID}
+	var result struct {
+		Data Bytes `json:"data"`
+	}
+	if err := c.FactomdRequest("raw-data", params, &result); err != nil {
+		return err
+	}
+
+	if err := f.UnmarshalBinary(result.Data); err != nil {
+		return err
+	}
+
+	txid, err := f.ComputeTransactionID()
```

Comment:
Move this to UnmarshalBinary. Populate the txid if nil, otherwise validate it. Populate it prior to Unmarshaling to then ensure that it is validated.